azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -86,12 +107,86 @@ public ProjectManager(ProjectLoader loader, Props props) {
     new XmlValidatorManager(prop);
     loadAllProjects();
     loadProjectWhiteList();
+
+    CleanerThread cleanerThread = new CleanerThread();
+    cleanerThread.start();
+  }
+
+  /**
+   * Cleaner thread to clean projects. Runs every day.
+  **/
+  private class CleanerThread extends Thread {
+
+    // runs every day
+    private static final long CLEANER_THREAD_WAIT_INTERVAL_MS = 24 * 60 * 60 * 1000;
+    private final User deleter = new User("azkaban");
+
+    private boolean shutdown = false;
+    private long lastProjectFileCleanTime = -1;
+
+    public CleanerThread() {
+      this.setName("ProjectManager-Cleaner-Thread");
+      logger.info("Starting cleaner thread: " + this.getName());
+    }
+
+    @SuppressWarnings("unused")
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
+    }
+
+    public void run() {
+      // There is no need for synchronized here as all the processing inside cleanProjectFilesAndInactivateProject
+      // is thread-safe and this is a single threaded process
+      while (!shutdown) {
+        try {
+            long currentTime = System.currentTimeMillis();
+
+            if (projectFilesCleanupEnabled &&
+                    (currentTime - CLEANER_THREAD_WAIT_INTERVAL_MS > lastProjectFileCleanTime)) {
+                logger.info("Cleaning up project files");
+                cleanProjectFilesAndInactivateProject();
+                lastProjectFileCleanTime = currentTime;
+            }
+
+            logger.info("Cleaner Thread completed. Sleeping!!!!!");
+            sleep(CLEANER_THREAD_WAIT_INTERVAL_MS);
+        } catch (ProjectManagerException e) {
+            logger.error(this.getName() + "failed due to: " + e.getMessage(), e);
+        } catch (Exception e) {
+            logger.error("Exception occurred. Probably to shut down.", e);
+        }
+      }
+    }
+
+    private void cleanProjectFilesAndInactivateProject() throws ProjectManagerException {
+      long currentTime = System.currentTimeMillis();
+
+      Set<Integer> scheduledProjectIds = projectLoader.fetchProjectIdsByEventType(EventType.SCHEDULE);
+      logger.info("Scheduled Projects count: " + scheduledProjectIds.size());
+
+      List<Project> projects = projectsById.values()
+              .stream()
+              .filter(project -> !scheduledProjectIds.contains(project.getId()) &&
+                      currentTime - project.getLastModifiedTimestamp() > projectFilesRetentionMs)
+              .collect(Collectors.toList());
+
+      logger.info("Project files to clean for: " + projects.size() + " projects");
+
+      for(Project project : projects) {
+        logger.info("Cleaning project files for project: " + project.getId());
+        projectLoader.cleanProjectFiles(project);
+        logger.info("Inactivating project and removing it from cache: " + project.getId());
```

Comment:
These calls internally are idempotent:

**Case1: If some or all project files are deleted, but projects are not removed, Azkaban shuts down.**
Next time when it comes back up, it will again check for projects modified before 15 days, then it will again find those projects. Some projects will have no files but still these projects will come, that time they will be marked inactive. Also, project files cleanup is a delete command in MySQL:
```
final String DELETE_PROJECT_FILES =
        "DELETE FROM project_files WHERE project_id=? AND version<?";
```
This is supposed to be idempotent

**Case2: If all project files are cleaned up but only some projects are deactivated, Azkaban shuts down.**
Next time when it comes back up, it will again check for projects modified before 15 days, then it will again find those projects. All projects will have no files but still these projects will come, that time they will be marked inactive.

Reply:
Can we document it in the code?

Reply:
added doc